### PR TITLE
make ops.add_foo take just class

### PR DESCRIPTION
**Pull request by jbrockmendel**
As discussed

Also small cleanup for Sparse ops.

### CODE REVIEW

File: pandas/core/ops.py [hunk 1]
```diff
@@ -567,6 +567,84 @@ def mask_cmp_op(x, y, op, allowed_types):
...
+
+    Parameters
+    ----------
+    cls : class
+
+    Returns
+    -------
+    arith_method : function
+    comp_method : function
+    bool_method : function
```
Comment (jreback) [inline pandas/core/ops.py]: I think it would be better just have a dict here keyed on the cls and just returns the 5 methods. This seems a bit overkill

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/ops.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/ops.py]: Sounds good.

Comment (jbrockmendel) [reply pandas/core/ops.py]: Implementing this it ends up being pretty close to a wash.  We have to key on `cls._typ` most of the time but also check for `cls._subtyp` for SparseSeries and SparseDataFrame.  At this point I'm leaning towards "being a little bit verbose here is OK in this case".  The main upside is its helpful to see which wrappers are used where without having to jump around the file.

Comment (jreback) [reply pandas/core/ops.py]: ok, can do in a single function then

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/ops.py [hunk 2]
```diff
@@ -788,22 +845,17 @@ def add_methods(cls, new_methods):
...
-        factory for special arithmetic methods:
-        f(cls, op, special)
-    comp_method : function (optional)
-        factory for rich comparison - signature: f(cls, op, special)
-    bool_method : function (optional)
-        factory for boolean methods - signature: f(cls, op, special)
+    cls : class
+        special methods will be defined and pinned to this class
     """
+    arith_method, comp_method, bool_method = _get_method_wrappers(cls)[2:]
```
Comment (jreback) [inline pandas/core/ops.py]: i'd rather you now do this an instead use ``_, _, ...`` for the return values its more obvious

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/ops.py [hunk 3]
```diff
@@ -1629,16 +1663,38 @@ def f(self, other, axis=0):
...
+    Returns
+    -------
+    left : SparseArray
+    right : SparseArray
+    """
+    opname = opname.strip('_')
+
+    if is_integer_dtype(left) and is_integer_dtype(right):
+        # series coerces to float64 if result should have NaN/inf
+        if opname in ('floordiv', 'mod') and (right.values == 0).any():
```
Comment (jreback) [inline pandas/core/ops.py]: this should go thru the fill_zeros path.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/ops.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/ops.py]: Is this obvious?  I don't know the sparse code all that well, have been assuming that the do-it-upfront approach taken here was for a reason.

Comment (jbrockmendel) [reply pandas/core/ops.py]: There will need to be a pass soon to make Series division handle division by zero the way that Index division now does.  That might be a reasonable time to take a close look at this.

Comment (jreback) [reply pandas/core/ops.py]: ok pls add it to the list.

**Review by jreback — APPROVED** [review]

**Pull request merged**